pkg/gardener/terraformer/terraformer.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package terraformer
+
+import (
+	gardenerterraformer "github.com/gardener/gardener/pkg/operation/terraformer"
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type terraformer struct {
+	tf *gardenerterraformer.Terraformer
+}
+
+// SetVariablesEnvironment implements Terraformer.
+func (t *terraformer) SetVariablesEnvironment(tfVarsEnvironment map[string]string) Interface {
+	return &terraformer{t.tf.SetVariablesEnvironment(tfVarsEnvironment)}
+}
+
+// InitializeWith implements Terraformer.
+func (t *terraformer) InitializeWith(initializer Initializer) Interface {
+	return &terraformer{t.tf.InitializeWith(initializer.Initialize)}
+}
+
+// Apply implements Terraformer.
+func (t *terraformer) Apply() error {
+	return t.tf.Apply()
+}
+
+// Destroy implements Terraformer.
+func (t *terraformer) Destroy() error {
+	return t.tf.Destroy()
+}
+
+// GetStateOutputVariables implements Terraformer.
+func (t *terraformer) GetStateOutputVariables(variables ...string) (map[string]string, error) {
+	return t.tf.GetStateOutputVariables(variables...)
+}
+
+// ConfigExists implements Terraformer.
+func (t *terraformer) ConfigExists() (bool, error) {
+	return t.tf.ConfigExists()
+}
+
+type initializer struct {
+	initialize func(config *gardenerterraformer.InitializerConfig) error
+}
+
+// Initialize implements Initializer.
+func (i *initializer) Initialize(config *gardenerterraformer.InitializerConfig) error {
+	return i.initialize(config)
+}
+
+type factory struct{}
```

Comment:
Since this interface has several methods, here using a `struct` is justified. However, at least the implementation could be simplified a lot by removing the `terraformer` and `initializer` types.

Reply:
As mentioned, `terraformer` cannot be removed, and for `Initializer` you'll still at least need the function adapter type. I've changed from the `initializer` struct to an `initializerFunc`, but I guess you probably can't do more about it.